File: UI/src/services/SettingsService.ts
import { apiAccess } from '@/services/ApiAccess';
import type { Settings } from "@/models/models";


class SettingsService {
    private _settings: Settings = {
        sortBy: "name",
        playbackSpeed: 1,
        sortDesc: true,
        preferredHeight: 720,
        useMDate: false,
        compatCodec: false,
        useNative: false
    }
    private _settingsPromise: Promise<Settings>;

    constructor(){
        this._settingsPromise = apiAccess.getSettings().then(z => {
            this._settings = z;
            return z;
        })
    }

    getSettingsAsync(): Promise<Settings>{
        return this._settingsPromise;
    }

    updatePlaybackSpeed(speed: number){
        if (this._settings.playbackSpeed != speed){
            this._settings.playbackSpeed = speed;
            apiAccess.updateSettings(this._settings);
        }
    }

    updateSort(sortBy: string, sortDesc: boolean){
        if (this._settings.sortBy != sortBy || this._settings.sortDesc != sortDesc){
            this._settings.sortBy = sortBy;
            this._settings.sortDesc = sortDesc;
            apiAccess.updateSettings(this._settings);
        }
    }

    updatePreferredHeight(preferredHeight: number | null){
        if (this._settings.preferredHeight != preferredHeight){
            this._settings.preferredHeight = preferredHeight;
            apiAccess.updateSettings(this._settings);
        }
    }

    updateUseMDate(useMDate: boolean){
        if (this._settings.useMDate != useMDate){            
            this._settings.useMDate = useMDate;
            apiAccess.updateSettings(this._settings);
        }
    }

    updateCompatCodec(compatCodec: boolean){
        if (this._settings.compatCodec != compatCodec){            
            this._settings.compatCodec = compatCodec;
            apiAccess.updateSettings(this._settings);
        }
    }

    updateUseNative(useNative: boolean){
        if (this._settings.useNative != useNative){            
            this._settings.useNative = useNative;
            apiAccess.updateSettings(this._settings);
        }
    }
}
export var settingsService = new SettingsService()

File: Server/Services/SettingsService.cs
using System.Collections.Concurrent;
using System.Net.Http.Json;
using System.Text.Json;

public class Settings
{
    public double PlaybackSpeed { get; set; } 
    public required string SortBy { get; set; }
    public bool SortDesc { get; set; }
    public int? PreferredHeight { get; set; }
    public bool UseMDate { get; set; }
    public bool CompatCodec { get; set; }
}

public class SettingsService
{
    private Settings _settings;
    private string _metadataPath = Path.Combine(BasePathHelper.BasePath, "metadata", "settings.txt");
    public SettingsService()
    {
        _settings = Load();
    }

    public Settings GetSettings()
    {
        return _settings;
    }

    public void UpdateSettings(Settings settings)
    {
        _settings = settings;
        string directoryPath = Path.GetDirectoryName(_metadataPath)!;
        if (!string.IsNullOrEmpty(directoryPath))
        {
            Directory.CreateDirectory(directoryPath);
        }
        string jsonContent = JsonSerializer.Serialize(_settings, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(_metadataPath, jsonContent);
    }

    private Settings Load()
    {
        if (File.Exists(_metadataPath))
        {
            try
            {
                var json = File.ReadAllText(_metadataPath);
                var settings = System.Text.Json.JsonSerializer.Deserialize<Settings>(json);
                if (settings != null)
                {
                    return settings;
                }
            }
            catch (Exception)
            {
                Console.WriteLine("error reading settings.txt");
            }
        }
        return new Settings
        {
            PlaybackSpeed = 1,
            SortBy = "name",
            SortDesc = true,
            PreferredHeight = 720,
            UseMDate = false,
            CompatCodec = false
        };
    }
}
